### ISSUE TITLE
Restrict duplicate bottom KDs for shelf stacks and other assemblies

### ISSUE THREAD

**Issue by TedCC1209** [CCDevNelson/Snap-Updates#111]
Can we add language to make it impossible to duplicate the bottom KD for shelf stacks (and for all assemblies I guess) when the bottom KD is already added in the partition prompts? There are three ways of getting that bottom KD that I can think of with shelf stacks and other assemblies. First, the opening sizes menu in the opening starter prompts:

![opening starter](https://user-images.githubusercontent.com/61025894/87080142-0a6b6e00-c1dc-11ea-810f-2c1ac3ade33e.png)

Construction options in the opening starter prompts:

![construction options](https://user-images.githubusercontent.com/61025894/87080267-3b4ba300-c1dc-11ea-89bb-a75e51e30bce.png)

And within the assembly prompt itself:

![assembly prompts](https://user-images.githubusercontent.com/61025894/87080400-75b54000-c1dc-11ea-878a-89e6bce9f65d.png)

I'm not sure how they all relate and talk to eachother, but we should write something to prevent duplicate bottom KDs.

**Comment by ClassyTeddy** [comment]
This is ready for testing on the `Restricting-All-Bottom-Duplicate-KD's` branch

**Comment by TEDSTANDAGE** [comment]
So, I'm not sure if there was an issue because this branch was created before the update, but I am still able to get doubled up bottom KDs on drawer stacks and hampers. However, it looks like shelf stacks are not doubling up, which was the main issue I posted about, so thank you for that. Also, it looks like the Doors are not doubling up (base and tall) either so thank you for that as well.

![image](https://user-images.githubusercontent.com/61025894/92311123-1f107c00-ef69-11ea-9621-40e13a4bc39a.png)

**Comment by ClassyTeddy** [comment]
I couldn't get a duplicate KD to appear for the hamper. What settings is it using?

**Comment by TEDSTANDAGE** [comment]
My guess is the KD's are coming from me having the down arrow here:

![Opening Sizes](https://user-images.githubusercontent.com/61025894/92506779-d2bc7a80-f1ba-11ea-89a4-538164bfcd18.png)

And then also the one that always comes in with Hampers. Here are the other prompts:

![Hamper Options](https://user-images.githubusercontent.com/61025894/92506817-e1a32d00-f1ba-11ea-85e0-b163d0633035.png)
![Construction Options](https://user-images.githubusercontent.com/61025894/92506820-e23bc380-f1ba-11ea-971f-31799c7f65ec.png)

**Comment by TEDSTANDAGE** [comment]
Testing the island fixes, I noticed that bottom KDs were being duplicated by the Hamper assembly there as well. Not sure how your restrictions are talking with/relating to the island partition prompts.

**Comment by ClassyTeddy** [comment]
I have a fix that is ready for testing on the `Restricting-All-Bottom-Duplicate-KD's` branch. Unfortunately, all these fixes/changes currently only with the hanging openings. What I did was change how hanging openings interacted with their inserts in order to avoid duplicates. I'll have to do a similar thing for islands. Luckily for islands it will be easier, because Islands cannot have their top and bottom KD's turned on and off.

**Comment by TEDSTANDAGE** [comment]
Looks good, thank you Teddy. 

I was able to break this by suspending some drawers, selecting the bottom KD option in the drawer prompts (which you have the user restricted from doing when they were not suspended), and then when I unselected the suspend drawers option the bottom KD option remained selected in the drawer prompts so it doubled up the bottom KD. Honestly though I had to try to break it so unless it's an easy fix I am fine moving forward with your fix as-is.

**Comment by ClassyTeddy** [comment]
It might be an easy fix, but the user should be able to fix that if they notice it by raising the drawer stack and removing the KD

**Comment by TEDSTANDAGE** [comment]
> It might be an easy fix, but the user should be able to fix that if they notice it by raising the drawer stack and removing the KD

Yeah, the user could fix a lot/most/if not all of these things by themselves if they know what to look for, but an average designer is not going to be looking for or noticing things like that. This fix is so there is less that engineering is going to have to change/alter/look for when doing a final review of a designer's drawing. I'm trying to put myself and others out of our engineering jobs as much as possible, letting SNaP do the work.

**Comment by ClassyTeddy** [comment]
I'll go and take a look at it once I finished all this flat crown and crown molding stuff

**Comment by ClassyTeddy** [comment]
Ready for testing on the `Restricting-All-Bottom-Duplicate-KD's`

**Comment by TEDSTANDAGE** [comment]
I still was able to break it with the suspended drawer situation but honestly let's just not worry about because it is such an unlikely scenario. It likely won't happen often and if it does happen they may just end up with one extra KD shelf if someone doesn't catch it when they look at the 2D views and see the double KD label there.

**Comment by TedCC1209** [comment]
Getting double bottom KDs again on shelf stacks in the master branch when the V (down) arrow is selected in the Opening Sizes prompts in the Opening Starting Prompts menu (to have the hanging assembly be floor based). Not sure if we want to reopen this issue or create a new one @TylerStandage

**Comment by TylerStandage** [comment]
Let's wait on this one to see if it's an issue in 2.0 testing version as it should have been fixed in 1.2.4. Will move this to needs triage so it doesn't get forgotten.

**Comment by TedCC1209** [comment]
Had a job go through the shop with duplicate bottom KDs (tall doors) that I didn't catch (v1.2.4). Here is the file:

https://drive.google.com/drive/folders/1-OBBvidGAo1_QrEIei__5grnDc61XBS-?usp=sharing

Would really like to get this one figured out.
